tracing/grpc.go

```diff
@@ -0,0 +1,188 @@
+package tracing
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/logging/logrus/ctxlogrus"
+	"go.opencensus.io/plugin/ocgrpc"
+	"go.opencensus.io/trace"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats"
+)
+
+//GRPCOption allows extending handler with additional functionality
+type GRPCOption func(*gRPCOptions)
+
+type metadataMatcher func(string) (string, bool)
+
+type gRPCOptions struct {
+	spanWithMetadata func(context.Context, stats.RPCStats) bool
+	metadataMatcher  metadataMatcher
+
+	spanWithPayload func(context.Context, stats.RPCStats) bool
+	maxPayloadSize  int
+}
+
+func defaultGRPCOptions() *gRPCOptions {
+	return &gRPCOptions{
+		metadataMatcher: defaultMetadataMatcher,
+		maxPayloadSize:  DefaultMaxPayloadSize,
+
+		//Keep spanWithMetadata and spanWithPayload equals to nil instead of dummy functions
+		//to prevent path trough header for each request
+	}
+}
+
+//WithMetadataAnnotation annotate span with request metadata
+func WithMetadataAnnotation(f func(context.Context, stats.RPCStats) bool) GRPCOption {
+	return func(ops *gRPCOptions) {
+		ops.spanWithMetadata = f
+	}
+}
+
+//WithMetadataMatcher set metadata matcher to filterout or preprocess metadata
+func WithMetadataMatcher(f func(string) (string, bool)) GRPCOption {
+	return func(ops *gRPCOptions) {
+		ops.metadataMatcher = f
+	}
+}
+
+//WithGRPCPayloadAnnotation add Inbound/Outbound payload as an attribute to span if f returns true
+func WithGRPCPayloadAnnotation(f func(context.Context, stats.RPCStats) bool) GRPCOption {
+	return func(ops *gRPCOptions) {
+		ops.spanWithPayload = f
+	}
+}
+
+//WithGRPCPayloadLimit limit payload size propogated to span
+//in case payload exceeds limit, payload truncated and
+//annotation payload.truncated=true added into span
+func WithGRPCPayloadLimit(limit int) GRPCOption {
+	return func(ops *gRPCOptions) {
+		ops.maxPayloadSize = limit
+	}
+}
+
+//Check that &ServerHandler{} comply stats.Handler interface
+var _ stats.Handler = &ServerHandler{}
+
+//ServerHandler is a wrapper over ocgrpc.ServerHandler
+//wrapper extends metadata added into the  span
+type ServerHandler struct {
+	ocgrpc.ServerHandler
+
+	options *gRPCOptions
+}
+
+//NewServerHandler returns wrapper over ocgrpc.ServerHandler
+func NewServerHandler(ops ...GRPCOption) *ServerHandler {
+	options := defaultGRPCOptions()
+	for _, op := range ops {
+		op(options)
+	}
+
+	return &ServerHandler{options: options}
+}
```

Comment:
How do I read the span from the context.Context passed into middleware? For example, authZ reads a span from the passed context to its middleware then creates additional spans inside of it to trace OPA. See here: https://github.com/Infoblox-CTO/ngp.authz/pull/745/files#diff-fd5b75f59732938618e94d5a57a06c35R112

Reply:
There are helper methods in this toolkit that take care of the plumbing for you.  You just request a span passing in the context.